interfaces/builtin/docker_support.go

```diff
@@ -60,6 +60,11 @@ const dockerSupportConnectedPlugAppArmor = `
 /{,var/}run/runc/       rw,
 /{,var/}run/runc/**     mrwklix,
 
+# Socket for docker-container-shim
+unix (bind,listen) type=stream addr="@/containerd-shim/moby/*/shim.sock\x00",
```

Comment:
I'm surprised by the NIL byte at the end, is that necessary?

Reply:
Yes, I can't understand why, but if it's not there apparmor won't allow it. If you want to test, try building my snapcraft.yaml for 18.05 from https://github.com/anonymouse64/docker-snap/tree/18.05

Reply:
Can you post the denial? Looking at https://github.com/containerd/containerd/blob/master/runtime/shim/client/client.go and https://github.com/containerd/containerd/blob/00d491063326ae7c173c9be3ba7bbd86c13d24c3/runtime/linux/bundle.go I don't see how the trailing nil byte happens to be added. The address is prefixed with one, but that's just denoting abstract socket, (which is `@..` in your case anyway).

Reply:
This is the denial itself:
```
[85046.199756] audit: type=1400 audit(1528721351.405:2267): apparmor="DENIED" operation="bind" profile="snap.docker.dockerd" pid=23360 comm="docker-containe" family="unix" sock_type="stream" protocol=0 requested_mask="bind" denied_mask="bind" addr="@2F636F6E7461696E6572642D7368696D2F6D6F62792F346435373936666261613265663435383432643733363131626433336637383032646163643934323461396533623130363536326263633039316335616330382F7368696D2E736F636B00"
```
Note that `aa-decode` complains about there being a null byte at the end (the `00` at the end of the address):
```
ijohnson@ijohnson-XPS-15-9560:~/git/docker-snap$ aa-decode 2F636F6E7461696E6572642D7368696D2F6D6F62792F346435373936666261613265663435383432643733363131626433336637383032646163643934323461396533623130363536326263633039316335616330382F7368696D2E736F636B00
/usr/sbin/aa-decode: line 40: warning: command substitution: ignored null byte in input
Decoded: /containerd-shim/moby/4d5796fbaa2ef45842d73611bd33f7802dacd9424a9e3b106562bcc091c5ac08/shim.sock
```
Here is how I reproduce this (I'm on 18.04 if that matters):
```
ijohnson@ijohnson-XPS-15-9560:~/git/docker-snap$ snap version
snap    2.32.9+18.04
snapd   2.32.8+18.04
series  16
ubuntu  18.04
kernel  4.15.0-22-generic
ijohnson@ijohnson-XPS-15-9560:~/git/docker-snap$ sudo snap install --jailmode --dangerous docker_18.05.0-ce_amd64.snap 
[sudo] password for ijohnson: 
docker 18.05.0-ce installed
ijohnson@ijohnson-XPS-15-9560:~/git/docker-snap$ sudo snap connect docker:docker-support
ijohnson@ijohnson-XPS-15-9560:~/git/docker-snap$ sudo snap connect docker:firewall-control 
ijohnson@ijohnson-XPS-15-9560:~/git/docker-snap$ sudo snap connect docker:privileged 
ijohnson@ijohnson-XPS-15-9560:~/git/docker-snap$ cat /var/lib/snapd/apparmor/profiles/snap.docker.dockerd | grep unix
  unix peer=(label=snap.@{SNAP_NAME}.*),
unix (bind) type=stream addr="@xtables",
ijohnson@ijohnson-XPS-15-9560:~/git/docker-snap$ sudo snap restart docker
Restarted.
ijohnson@ijohnson-XPS-15-9560:~/git/docker-snap$ docker run -it ubuntu bash
Unable to find image 'ubuntu:latest' locally
latest: Pulling from library/ubuntu
6b98dfc16071: Pull complete 
4001a1209541: Pull complete 
6319fc68c576: Pull complete 
b24603670dc3: Pull complete 
97f170c87c6f: Pull complete 
Digest: sha256:5f4bdc3467537cbbe563e80db2c3ec95d548a9145d64453b06939c4592d67b6d
Status: Downloaded newer image for ubuntu:latest
docker: Error response from daemon: failed to listen to abstract unix socket "/containerd-shim/moby/4d5796fbaa2ef45842d73611bd33f7802dacd9424a9e3b106562bcc091c5ac08/shim.sock": listen unix /containerd-shim/moby/4d5796fbaa2ef45842d73611bd33f7802dacd9424a9e3b106562bcc091c5ac08/shim.sock: bind: permission denied: unknown.
ijohnson@ijohnson-XPS-15-9560:~/git/docker-snap$ dmesg | grep DENIED | tail -n 1
[85046.199756] audit: type=1400 audit(1528721351.405:2267): apparmor="DENIED" operation="bind" profile="snap.docker.dockerd" pid=23360 comm="docker-containe" family="unix" sock_type="stream" protocol=0 requested_mask="bind" denied_mask="bind" addr="@2F636F6E7461696E6572642D7368696D2F6D6F62792F346435373936666261613265663435383432643733363131626433336637383032646163643934323461396533623130363536326263633039316335616330382F7368696D2E736F636B00"
ijohnson@ijohnson-XPS-15-9560:~/git/docker-snap$ aa-decode 2F636F6E7461696E6572642D7368696D2F6D6F62792F346435373936666261613265663435383432643733363131626433336637383032646163643934323461396533623130363536326263633039316335616330382F7368696D2E736F636B00
/usr/sbin/aa-decode: line 40: warning: command substitution: ignored null byte in input
Decoded: /containerd-shim/moby/4d5796fbaa2ef45842d73611bd33f7802dacd9424a9e3b106562bcc091c5ac08/shim.sock
ijohnson@ijohnson-XPS-15-9560:~/git/docker-snap$ sudo gedit /var/lib/snapd/apparmor/profiles/snap.docker.dockerd 
ijohnson@ijohnson-XPS-15-9560:~/git/docker-snap$ cat /var/lib/snapd/apparmor/profiles/snap.docker.dockerd | grep unix
  unix peer=(label=snap.@{SNAP_NAME}.*),
unix (bind) type=stream addr="@xtables",
unix (bind,listen) type=stream addr="@/containerd-shim/moby/*/shim.sock",
ijohnson@ijohnson-XPS-15-9560:~/git/docker-snap$ sudo apparmor_parser -r /var/lib/snapd/apparmor/profiles/snap.docker.dockerd 
ijohnson@ijohnson-XPS-15-9560:~/git/docker-snap$ journalctl -e | grep apparmor_parser | tail -n 1
Jun 11 07:51:59 ijohnson-XPS-15-9560 kernel: audit: type=1400 audit(1528721519.119:2268): apparmor="STATUS" operation="profile_replace" profile="unconfined" name="snap.docker.dockerd" pid=24068 comm="apparmor_parser"
ijohnson@ijohnson-XPS-15-9560:~/git/docker-snap$ sudo snap restart docker
Restarted.
ijohnson@ijohnson-XPS-15-9560:~/git/docker-snap$ docker run -it ubuntu bash
docker: Error response from daemon: failed to listen to abstract unix socket "/containerd-shim/moby/7d2bee427f79966ee848fa37afedb821c23c8b135d6f2252bd4cfaa962adf20d/shim.sock": listen unix /containerd-shim/moby/7d2bee427f79966ee848fa37afedb821c23c8b135d6f2252bd4cfaa962adf20d/shim.sock: bind: permission denied: unknown.
ijohnson@ijohnson-XPS-15-9560:~/git/docker-snap$ dmesg | grep DENIED | tail -n 1
[85333.372882] audit: type=1400 audit(1528721638.577:2285): apparmor="DENIED" operation="bind" profile="snap.docker.dockerd" pid=24327 comm="docker-containe" family="unix" sock_type="stream" protocol=0 requested_mask="bind" denied_mask="bind" addr="@2F636F6E7461696E6572642D7368696D2F6D6F62792F376432626565343237663739393636656538343866613337616665646238323163323363386231333564366632323532626434636661613936326164663230642F7368696D2E736F636B00"
ijohnson@ijohnson-XPS-15-9560:~/git/docker-snap$ aa-decode 2F636F6E7461696E6572642D7368696D2F6D6F62792F376432626565343237663739393636656538343866613337616665646238323163323363386231333564366632323532626434636661613936326164663230642F7368696D2E736F636B00
/usr/sbin/aa-decode: line 40: warning: command substitution: ignored null byte in input
Decoded: /containerd-shim/moby/7d2bee427f79966ee848fa37afedb821c23c8b135d6f2252bd4cfaa962adf20d/shim.sock
ijohnson@ijohnson-XPS-15-9560:~/git/docker-snap$ sudo gedit /var/lib/snapd/apparmor/profiles/snap.docker.dockerd 
ijohnson@ijohnson-XPS-15-9560:~/git/docker-snap$ sudo apparmor_parser -r /var/lib/snapd/apparmor/profiles/snap.docker.dockerd 
ijohnson@ijohnson-XPS-15-9560:~/git/docker-snap$ journalctl -e | grep apparmor_parser | tail -n 1
Jun 11 07:55:57 ijohnson-XPS-15-9560 kernel: audit: type=1400 audit(1528721757.051:2303): apparmor="STATUS" operation="profile_replace" profile="unconfined" name="snap.docker.dockerd" pid=25036 comm="apparmor_parser"
ijohnson@ijohnson-XPS-15-9560:~/git/docker-snap$ sudo snap restart docker
Restarted.
ijohnson@ijohnson-XPS-15-9560:~/git/docker-snap$ docker run -it ubuntu bash
root@d9dde11121e5:/# exit
ijohnson@ijohnson-XPS-15-9560:~/git/docker-snap$ 
```

Reply:
The NUL byte is needed because it is what docker (at least in this snap) is putting at the end:
```
addr="@...00"
```

NUL bytes are allowed in abstrack socket addresses. I'm more concerned with 'moby' in the path. Is this a codename? A container name? Something else that may change? If so, perhaps this rule would be better:
```
unix (bind,listen) type=stream addr="@/containerd-shim/**/shim.sock\x00",
```

Reply:
`moby` here is the new name of the docker project (i.e. see this [post](https://blog.docker.com/2017/04/introducing-the-moby-project/) and the fact that the old [github.com/docker/docker](https://github.com/docker/docker) is now [github.com/moby/moby](https://github.com/moby/moby) ). I would say it's unlikely to change, but since docker changed to moby I guess it is possible it could be changed to something else. I can change if if y'all think it should be changed, but I'm leaning towards leaving it as is with moby there.

Reply:
Oh! I totally missed that. If it is the project name, then I think it is a *good* thing to have it in the path-- it will separate it from other implementations that use containerd-shim with a similar path.